src/common/Utils.ts

```diff
@@ -104,21 +105,42 @@ export async function getEip712Signature (
   })
 }
 
+/**
+ * @returns the actual cost of putting this transaction on chain.
+ */
+export function calculateCalldataCost (calldata: string): number {
+  const calldataBuf = Buffer.from(calldata.replace('0x', ''), 'hex')
+  let sum = 0
+  calldataBuf.forEach(ch => { sum += (ch === 0 ? defaultEnvironment.gtxdatazero : defaultEnvironment.gtxdatanonzero) })
+  return sum
+}
+
 /**
  * @returns maximum possible gas consumption by this relayed call
+ * (calculated on chain by RelayHub.verifyGasAndDataLimits)
  */
 export function calculateTransactionMaxPossibleGas (
   {
     gasAndDataLimits,
     hubOverhead,
     relayCallGasLimit,
-    msgDataGasCost
+    msgDataGasCost,
+    txDataCost
   }: TransactionGasCostComponents): number {
   return hubOverhead +
-    msgDataGasCost +
-    parseInt(relayCallGasLimit) +
-    parseInt(gasAndDataLimits.preRelayedCallGasLimit) +
-    parseInt(gasAndDataLimits.postRelayedCallGasLimit)
+      parseInt(gasAndDataLimits.preRelayedCallGasLimit) +
+      parseInt(gasAndDataLimits.postRelayedCallGasLimit) +
+      parseInt(relayCallGasLimit) +
+      msgDataGasCost +
+      txDataCost
+
+  // maxPossibleGas =
+  //     gasOverhead.add(
+  //         gasAndDataLimits.preRelayedCallGasLimit).add(
+  //         gasAndDataLimits.postRelayedCallGasLimit).add(
+  //         relayRequest.request.gas).add(
+  //         dataGasCost).add(
+  //         txDataCost);
```

Comment:
\--

Reply:
ok (though this line is a match to a solidity code too bad both use some unreadable construct - and each is different...